### ISSUE TITLE
cl19, the 1km OS grid spatial layer, is not being sampled

### ISSUE THREAD

**Issue by sophiathirza** [nbnuk/nbnatlas-issues#680]
cl19 has stopped being sampled, which I think is causing the problem that it's not possible to view the records in 1km qrid squares by clicking on a square in the interactive map. Also, I've sent quite a few people an example url to search for records by 1km GR and this is no longer working.

**Comment by reupost** [comment]
I've reloaded this layer as no. 149 and added the metadata from the old version. This new layer is available on the spatial portals and has registered successfully for sampling as well. I couldn't figure out what was wrong with the old one. Unfortunately it means any queries for 'cl19=xxx' will need to be rewritten as 'cl149=xxx'.

**Comment by sophiathirza** [comment]
no worries, that's great - thanks.

**Comment by sophiathirza** [comment]
The sampling hasn't worked for the 1km grid: https://records.nbnatlas.org/occurrences/search?q=cl149%3A%22HP4001%22&fq=geospatial_kosher%3A*&fq=*%3A*+AND+-%28occurrence_status%3A%22absent%22%29#tab_recordsView

there are records all over the UK in this set

**Comment by reupost** [comment]
That's true. Would you mind checking that we used the correct unique ID field? And that it was actually unique? I've never seen anything like this - that's the only explanation I can think of.

**Comment by sophiathirza** [comment]
There is only one attribute: PLAN_NO

![image](https://user-images.githubusercontent.com/1651015/55064793-e2a1d680-507a-11e9-93fc-ec8e24cbcaa4.png)

can PLAN_NO be used for both the ID and the Name?

**Comment by sophiathirza** [comment]
There is a gap in the grids: 
![image](https://user-images.githubusercontent.com/1651015/55065373-0285ca00-507c-11e9-833f-44d06875fd84.png)

It was reported a few months ago on the source github site. Is it something that we could correct?

**Comment by reupost** [comment]
We did use PLAN_NO for both ID and name field. I wouldn't have thought that would cause a problem, assuming they are unique.

Possibly there is another problem with the shapefile, apart from simply the gap. Possibly Christine could fill that in for us, but the scatter of records outside of a named grid cell implies there might be a problem with the topology as well.

**Comment by sophiathirza** [comment]
Yes, PLAN_NO is unique.

@CEJohnston would you be able to look at the 1km shape file to see if you can fill in the missing gap and see whether there are any problems with the topology? thanks

**Comment by CEJohnston** [comment]
Have checked the topology (my file was downloaded from source GitHub page) and no errors reported.  I can't think of any reason why the search for HP4001 behaves the way it does.  I'm afraid I am not sure how to fill the gap with precise 1km squares - I think it might need scripting to do properly and quickly.  However, even if it was filled I am not sure if it would fix the layer.
Is it a problem that the layer is only complete for terrestrial and not marine areas that are within OSGB? The gap is in the Firth of Clyde because the 1k squares have been clipped to the coastline + a small buffer.  The same is true for the 10k and 100k grids.  
Image shows the 100k grid over the 1k grid.  Completely missing are 100k squares HO, NE and NP (all Scotland) but 1k square coverage (red) within for example NB, NG and NL (Outer Hebrides) is patchy and not covering coastal waters.
![image](https://user-images.githubusercontent.com/18594564/55096223-5fe83e00-50b1-11e9-9fef-08a2d4fca49a.png)
Ideally to make the 1k grid usable for all possible grid reference search options it needs to be extended out to the edge of the full 100k grid (although if the search is done through the spatial portal it is obvious that only selected, largely terrestrial grids, can be defined).  Not quite sure what to suggest next.

**Comment by CEJohnston** [comment]
Having looked at creating a new grid that extends out to the full 100k coverage I can't find a way of assigning latitudes and longitudes to the grid squares and hence to generate the OSGB references.  The usual tools to generate lats and longs return null values.  The geometry of the polygons appears to be defined relative to the origin of the grid (in metres from that point) and not to a coordinate reference system such as WGS84.  This may be a reason for the original 1k grid not sampling.  It needs further investigation if the grids are to be extended.

**Comment by sophiathirza** [comment]
Coming back to sampling the 1km grid:

The following seems to work:
https://records.nbnatlas.org/occurrences/search?q=*%3A*&fq=grid_ref_1000%3ASO5214

What is the difference between using cl149 and grid_ref_1000? I assume that grid_ref_1000 doesn't use the shape file but perhaps it only works for those records supplied with a grid reference?

**Comment by CEJohnston** [comment]
Would it be able to calculate from the record's actual grid ref? I don't know what the difference is between the two parts of the code - is there something else in the database it can call on?

On request the Ordnance Survey have just sent me shapefiles of grids for 1, 5, 10, 20, and100km but they cover almost the same extents as our existing ones (ie they are just for terrestrial areas and contain the same gaps in the inshore water areas).  I can upload if you think worth testing?

Will follow up with the UK Hydrographic Office to see if they have grids for sea areas.

**Comment by CEJohnston** [comment]
Response received from the UK Hydrographic Office: they do not have any grids that cover the marine and inshore water areas of the OSGB.

**Comment by CEJohnston** [comment]
I have uploaded the 1km grid that I received from the Ordnance Survey for testing in the portal.

**Comment by sophiathirza** [comment]
Zip file name: 1km_Grid_Shape.zip
Shape file name: 1KGD99_region.shp
Layer name: Ordnance Survey Grids (OSGB) - 1km
ID field: PLAN_NO
Name field: PLAN_NO

**Comment by reupost** [comment]
Loaded as layer 167. Please add metadata as needed. If it samples correctly I will use it to replace the old layer, if that is appropriate?

**Comment by CEJohnston** [comment]
The layer metadata has been completed as far as possible.  There is no download link as the file was supplied by email.  Have not been able to test the sampling as the search is timing out.  Might be worth testing with the known grid references that failed previously.

**Comment by CEJohnston** [comment]
Just noting an update from Sophie (received by email when I was reviewing the spatial layer issues but worth sharing here too):  this one is still an issue but possibly we don't need it if we use the grid reference determined by the code rather than the spatial layer. There is also this outstanding question of putting all our records on a single grid.
